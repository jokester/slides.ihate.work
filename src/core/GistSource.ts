import { Octokit } from '@octokit/rest';
import { SourceBundle } from './SourceBundle';
import { SourceError } from './errors';

interface GistSourceLocator {
  ownerId: string;
  gistId: string;
  revisionId?: string;
}

export class GistSource {
  /**
   * @param gistUrl like:
   * - https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8
   * => page: gist
   * - https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/4676f49e32f95fd76549ce4f7330f0f7aa4662b3
   * => page: gist revision
   * - https://gist.githubusercontent.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/raw/4676f49e32f95fd76549ce4f7330f0f7aa4662b3/0-rancher-zerotier-k3s-deployment.md
   * => an raw file
   * @param client
   */
  constructor(
    readonly gistUrl: string,
    private readonly client = new Octokit(),
  ) {}

  get fetchKey(): string {
    return `gistId=${this.gistUrl}`;
  }

  get pageUrl(): string | null {
    const parsed = this.parseUrl();
    if (!parsed) {
      return null;
    }

    return `https://gist.github.com/${parsed.ownerId}/${parsed.gistId}`;
  }

  parseUrl(): GistSourceLocator | null {
    const u = new URL(this.gistUrl);
    const parts = u.pathname.split('/');
    const [_empty, ownerId, gistId, revisionId, ...rest] = parts;
    if (gistId?.length === 32 && !revisionId) {
      return {
        ownerId,
        gistId,
      };
    }
    // TODO: support more patterns
    return null;
  }

  /**
   * @return a asset bundle fetched from github
   */
  async fetchSource(): Promise<SourceBundle> {
    const res = await this.client.rest.gists.get({
      gist_id: this.parseUrl()!.gistId,
    });

    const files = Object.values(res.data.files || {});
    const mds = files.filter((file) => file?.filename?.toLowerCase().endsWith('.md'));
    if (!mds.length) {
      throw new SourceError('no markdown file found in gist');
    }
    if (mds.length > 1) {
      throw new SourceError('multiple markdown files found in gist');
    }

    const [mdFile] = mds;

    // TODO: fetch assets (most likely to be images) from gist files or gist comments?
    return {
      slideText: mdFile?.content ?? '',
    };
  }
}
