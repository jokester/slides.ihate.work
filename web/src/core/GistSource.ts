import { Octokit } from '@octokit/rest';
import { SlideBundle } from './SlideBundle';
import { SourceError } from './errors';
import { isUrl } from './url-loader';
import { InternalUrlProvider } from './internal-url-provider';

interface GistSourceLocator {
  rawUrl: string;
  ownerId: string;
  gistId: string;
  revisionId?: string;
}

export function buildGistSource(url: URL): GistSource | null {
  const l = parseGistUrl(url);
  return l && new GistSource(l);
}

/**
 * - https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8
 * => page: gist
 * - https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/4676f49e32f95fd76549ce4f7330f0f7aa4662b3
 * => page: gist revision
 * - https://gist.githubusercontent.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/raw/4676f49e32f95fd76549ce4f7330f0f7aa4662b3/0-rancher-zerotier-k3s-deployment.md
 * => an raw file
 */
export function parseGistUrl(u: URL): GistSourceLocator | null {
  const parts = u.pathname.split('/');
  const [_empty, ownerId, gistId, revisionId, ...rest] = parts;
  if (gistId?.length === 32 && !revisionId) {
    return {
      rawUrl: u.toString(),
      ownerId,
      gistId,
    };
  }
  // TODO: support more patterns
  return null;
}

class GistSource implements InternalUrlProvider {
  constructor(
    readonly locator: Readonly<GistSourceLocator>,
    private readonly client = new Octokit(),
  ) {}

  get fetchKey(): string {
    return `gistUrl=${this.locator.rawUrl}`;
  }

  asInternalPageUrl(): string {
    const parsed = this.locator;
    return `/gist/${parsed.ownerId}/${parsed.gistId}`;
  }

  asUpstreamUrl(): string {
    return this.locator.rawUrl;
  }

  /**
   * @return a asset bundle fetched from GitHub
   */
  async fetchSource(): Promise<SlideBundle> {
    const res = await this.client.rest.gists.get({
      gist_id: this.locator.gistId,
    });

    const files = Object.values(res.data.files || {});
    const mds = files.filter((file) => file?.filename?.toLowerCase().endsWith('.md'));
    if (!mds.length) {
      throw new SourceError('no markdown file found in gist');
    }
    if (mds.length > 1) {
      throw new SourceError('multiple markdown files found in gist');
    }

    const [mdFile] = mds;

    // TODO: fetch assets (most likely to be images) from gist files or gist comments?
    return {
      slideText: mdFile?.content ?? '',
      gistSource: this,
    };
  }
}
