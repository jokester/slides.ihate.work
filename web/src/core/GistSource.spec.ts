import { parseGistUrl } from './GistSource';

describe('GistSource', () => {
  describe('parseGistUrl', () => {
    it('should parse gist url', () => {
      for (const url of [
        'https://gist.githubusercontent.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/raw/4676f49e32f95fd76549ce4f7330f0f7aa4662b3/0-rancher-zerotier-k3s-deployment.md',
        'https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8',
        'https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/4676f49e32f95fd76549ce4f7330f0f7aa4662b3',
      ]) {
        const parsed = parseGistUrl(new URL(url));
        expect(parsed).toBeTruthy();
        expect(parsed).toMatchSnapshot(`parseGistUrl(${url})`);
      }
    });
  });
});
