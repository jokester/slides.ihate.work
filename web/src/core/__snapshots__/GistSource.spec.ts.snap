// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GistSource parseGistUrl should parse gist url: parseGistUrl(https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8) 1`] = `
{
  "gistId": "2ae304016d8c25b09a68a9221f6c07c8",
  "ownerId": "jokester",
  "rawUrl": "https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8",
}
`;

exports[`GistSource parseGistUrl should parse gist url: parseGistUrl(https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/4676f49e32f95fd76549ce4f7330f0f7aa4662b3) 1`] = `
{
  "gistId": "2ae304016d8c25b09a68a9221f6c07c8",
  "ownerId": "jokester",
  "rawUrl": "https://gist.github.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/4676f49e32f95fd76549ce4f7330f0f7aa4662b3",
  "revisionId": "4676f49e32f95fd76549ce4f7330f0f7aa4662b3",
}
`;

exports[`GistSource parseGistUrl should parse gist url: parseGistUrl(https://gist.githubusercontent.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/raw/4676f49e32f95fd76549ce4f7330f0f7aa4662b3/0-rancher-zerotier-k3s-deployment.md) 1`] = `
{
  "filename": "0-rancher-zerotier-k3s-deployment.md",
  "gistId": "2ae304016d8c25b09a68a9221f6c07c8",
  "ownerId": "jokester",
  "rawUrl": "https://gist.githubusercontent.com/jokester/2ae304016d8c25b09a68a9221f6c07c8/raw/4676f49e32f95fd76549ce4f7330f0f7aa4662b3/0-rancher-zerotier-k3s-deployment.md",
  "revisionId": "4676f49e32f95fd76549ce4f7330f0f7aa4662b3",
}
`;
